### PR TITLE
impl(auth): get_headers for user credentials

### PR DESCRIPTION
**Pull request by dbolduc**
Part of the work for #442 

Implement the `get_headers()` function (based on the `get_token()` function) for user credentials.

Note that we set the header as sensitive, which is what `reqwest::RequestBuilder::bearer_auth(...)` does: https://docs.rs/reqwest/0.12.9/src/reqwest/async_impl/request.rs.html#270. (I could not find a public ctor that took `is_sensitive`, so we do it in two lines).

---

Is there a better way to set test expectations on a container?

TIL: `map_err()`. Check me on my error propagation.

### CODE REVIEW

File: src/auth/src/credentials/user_credential.rs [hunk 1]
```diff
@@ -37,8 +37,12 @@ where
...
 
     async fn get_headers(&mut self) -> Result<Vec<(HeaderName, HeaderValue)>> {
-        // TODO(#442) - implementation in progress
-        Err(CredentialError::new(false, BoxError::from("unimplemented")))
+        let token = self.get_token().await?;
+        let mut value = HeaderValue::from_str(&format!("{} {}", token.token_type, token.token))
+            .map_err(|e| CredentialError::new(false, e.into()))?;
+        value.set_sensitive(true);
+        let headers = vec![(AUTHORIZATION, value)];
+        Ok(headers)
```
Comment (coryan) [inline src/auth/src/credentials/user_credential.rs:45]: I would probably just write:

```suggestion
        Ok(vec![(AUTHORIZATION, value)])
```

But both seem fine.

File: src/auth/src/credentials/user_credential.rs [hunk 2]
```diff
@@ -78,11 +82,46 @@ mod test {
...
+        mock.expect_get_token().times(1).return_once(|| Ok(token));
+
+        let mut uc = UserCredential {
+            token_provider: mock,
+        };
+        let actual = uc.get_headers().await.unwrap();
+        assert_eq!(actual.len(), 1);
+        assert_eq!(actual[0].0, AUTHORIZATION);
+        assert_eq!(actual[0].1, "Bearer test-token");
+        assert!(actual[0].1.is_sensitive());
```
Comment (coryan) [inline src/auth/src/credentials/user_credential.rs:112]: Consider something like:

https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=af731026c583586602b056c07e2b44b4

You may need to transform the `HeaderMap` into something with strings...

**Review by coryan — APPROVED** [review]
The tests could use some cleanups, the rest is just nitpicky.

File: src/auth/src/credentials/user_credential.rs [hunk 3]
```diff
@@ -78,11 +82,46 @@ mod test {
...
+    }
+
+    #[tokio::test]
+    async fn get_headers_failure() {
+        let mut mock = MockTokenProvider::new();
+        mock.expect_get_token()
+            .times(1)
+            .return_once(|| Err(CredentialError::new(false, Box::from("fail"))));
```
Comment (sai-sunder-s) [inline src/auth/src/credentials/user_credential.rs:120]: consider adding a test where retryable is returned true from the provider.

File: src/auth/src/credentials/user_credential.rs (hunk 1 above)
Comment (dbolduc) [reply src/auth/src/credentials/user_credential.rs]: Fixed.

File: src/auth/src/credentials/user_credential.rs (hunk 2 above)
Comment (dbolduc) [reply src/auth/src/credentials/user_credential.rs]: > You may need to transform the `HeaderMap` into something with strings...

Done.

The conversion into the struct is a bit gross (aesthetically), but I find the code easy to read, so I am happy with it.

File: src/auth/src/credentials/user_credential.rs (hunk 3 above)
Comment (dbolduc) [reply src/auth/src/credentials/user_credential.rs:141]: No thanks. I am not too concerned about accidentally manipulating the error before returning it here:

https://github.com/googleapis/google-cloud-rust/blob/60bbda3220e13784278e6b33c286df3b2fa91122/src/auth/src/credentials/user_credential.rs#L40

**Pull request merged**